runtime/java/src/common/DecoratedNode.java

```diff
@@ -232,8 +232,10 @@ public DecoratedNode decorate(final DecoratedNode parent, final Lazy[] inhs, fin
 	}
 
 	private void copyInhOverrides(final DecoratedNode parent, final Lazy[] inhs, final Lazy[] newInhs) {
-		assert inhs.length == newInhs.length;
-		for(int i = 0; i < inhs.length; i++) {
+		// Arrays can differ in length if a tree is shared in a polymorphic decoration site.
+		// The new inh array should never be larger than the original one.
```

Comment:
You should probably comment here *why* the new one should never be larger than the original, unless this is absolutely clear for someone using this.